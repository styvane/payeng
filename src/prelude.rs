pub use crate::account::*;
pub use crate::client::Client;
pub use crate::result::Result;
pub use crate::transaction::*;
