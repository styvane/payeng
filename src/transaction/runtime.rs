//! Transaction processor.
//!

use std::fs::File;
use std::io;
use std::path::Path;
use std::thread;

use crossbeam::channel;

use super::{Reader, Writer};
use crate::error::Error;
use crate::transport::Sender;
use crate::Result;

/// Run everything.
#[tracing::instrument(name = "Run all", skip(reader, writer, capacity))]
pub fn run(
    reader: impl io::Read + Send + 'static,
    writer: impl io::Write + Send + 'static,
    capacity: usize,
) {
    let (outgoing, incoming) = channel::bounded(capacity);
    let mut reader = Reader::from_reader(reader, outgoing);
    let mut writer = Writer::from_writer(writer, incoming);

    let r_handle = thread::spawn(move || reader.send());
    let w_handle = thread::spawn(move || writer.write());

    if let Err(err) = r_handle.join() {
        tracing::error!(err.cause_chain=?err);
    }
    if let Err(err) = w_handle.join() {
        tracing::error!(err.cause_chain=?err);
    }
}

/// Creates an io::Reader from file path.
fn new_reader(path: impl AsRef<Path>) -> Result<impl io::Read> {
    File::open(path).map_err(Error::IoError)
}

/// Gets input file.

pub fn get_input() -> Result<impl io::Read> {
    let path = std::env::args().nth(1).ok_or(Error::InvalidArgumentError)?;
    new_reader(path)
}
