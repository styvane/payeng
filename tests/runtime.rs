use parking_lot::Mutex;
use payeng::prelude::runtime;
use std::sync::Arc;

struct TestWriter {
    content: Arc<Mutex<Vec<u8>>>,
}

impl std::io::Write for TestWriter {
    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
        self.content.lock().write(buf)
    }

    fn flush(&mut self) -> std::io::Result<()> {
        self.content.lock().flush()
    }
}

#[test]
fn run_output_expected_value() {
    let dir = std::env::current_dir().expect("failed to get current directory");
    let path = dir.join("tests/test.csv");
    let reader = std::fs::File::open(path).unwrap();
    let content = Arc::new(Mutex::new(vec![]));
    let writer = TestWriter {
        content: content.clone(),
    };
    runtime::run(reader, writer, 20);
    let content = content.lock().clone();
    let mut settings = insta::Settings::clone_current();
    settings.set_sort_maps(true);
    settings.bind(|| {
        insta::assert_csv_snapshot!(
            String::from_utf8(content).expect("failed to convert to string")
        );
    });
}
