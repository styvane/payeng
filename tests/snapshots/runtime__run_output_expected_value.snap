---
source: tests/runtime.rs
expression: "String::from_utf8(content).expect(\"failed to convert to string\")"
---
"client,available,held,total,locked
2,2,0,2,false
1,1.5,0,1.5,false
"
